udata/core/spam/models.py

```diff
@@ -0,0 +1,161 @@
+        """
+        callbacks = self.spam.callbacks
+        self.spam.status = NO_SPAM
+        self.spam.callbacks = {}
+        base_model.save()
+
+        for name, args in callbacks.items():
+            callback = getattr(base_model, name)
+            callback(*args['args'], **args['kwargs'])
+
+    def is_spam(self):
+        return self.spam and self.spam.status == POTENTIEL_SPAM
+    
+    def attributes_to_check_for_spam(self):
+        raise NotImplementedError("Please implement the `attributes_to_check_for_spam` method. Should return the list of attributes to check.")
+
+    def embeds_to_check_for_spam(self):
+        return []
+    
+    def spam_report_title(self):
+        return type(self).__name__
+    
+    def spam_report_link(self):
+        return None
+    
+    def _report(self, text, breadcrumb, reason):
+        # Note that all the chain should be a SpamMixin, maybe we could filter out if it's not the case here…
+        title = " → ".join(map(lambda o: o.spam_report_title(), breadcrumb))
+
+        # Select the first link in the embed list (for exemple message doesn't have a link, so check if discussion have one)
+        for object in reversed(breadcrumb):
+            link = object.spam_report_link()
+            if link:
+                break
+
+        on_new_potential_spam.send(title=title, link=link, text=text, reason=reason)
+
+
+def spam_protected(get_model_to_check=None):
+    """
+    This decorator prevent the calling of a class method if the object is a POTENTIAL_SPAM.
+    It will save the class method called with its arguments inside the `SpamInfo` object to be
+    called later if needed (in case of false positive).
+    The decorator accept an argument, a function to get the model to check when we are doing an operation
+    on an embed document. The class method should always take a `self` as a first argument which is the base 
+    model to allow saving the callbacks back into Mongo (we cannot .save() an embed document).
+    """
+    def decorator(f):
+        def protected_function(*args, **kwargs):
+            base_model = args[0]
+            if get_model_to_check:
+                model_to_check = get_model_to_check(*args, **kwargs)
+            else:
+                model_to_check = base_model
+
+            if not isinstance(model_to_check, SpamMixin):
+                raise ValueError("@spam_protected should be called within a SpamMixin. " + type(model_to_check).__name__ + " given.")
+
+            if model_to_check.is_spam():
+                model_to_check.spam.callbacks[f.__name__] = {
+                    'args': args[1:],
+                    'kwargs': kwargs
+                }
+                # At this point we recall `detect_spam`, but there is a check inside `detect_spam` to not do nothing if 
+                # we only change spam.* fields.
+                # Here we call save() on the base model because we cannot save an embed document.
```

Comment:
I thought about implementing a `save()` function on the embeds, we could use `resource._instance` to get the parent and call `resource._instance.save()` but it's not always working because `_instance` is a week pointer so if the parent goes out of  scope in the app, the `resource._instance` can become `null` (at least that's my understanding at this point…)

Reply:
Indeed! I'm not sure it can occur in this case, but we can have a fail-safe on this. See for example : https://github.com/opendatateam/udata/blob/95dd559fb984d5a9d705f035fb89b811fafaa146/udata/core/dataset/models.py#L442.

Reply:
I tried this solution but not sure how to make it work… Because if `_instance` is `None`, we'll query a new base model but I modified the embed outside this new queried base model. The `base_model.save()` is useless in this case. I need to modified the embed inside the new base model but it's really hard to get to the embed from the new one. I'm not sure I'm clear :-D

Reply:
Yes, I think I see what you mean, you would need to re-apply current modifications to one of the loaded embeds (which?).
If it gets too messy to implement a save method, we can keep it this way

Reply:
Yes I think it's too messy right now… i'm open to new ideas because my solution is far from ideal… Maybe we can keep it as it is and see if someone have a better idea in the futur to improve the code.